libslide/src/evaluator_rules.rs

```diff
@@ -109,36 +112,64 @@ impl Transformer<Expr, Expr> for BuiltRule {
     ///   "$x + 0 -> $x".try_apply("x + 1")  // None
     ///   "$x + 0 -> $x".try_apply("x")      // None
     ///
-    fn transform(&self, target: Expr) -> Expr {
-        // First, apply the rule recursively on the target's subexpressions.
-        let partially_transformed = match target {
-            expr @ Expr::Const(_) => expr,
-            var @ Expr::Var(_) => var,
-            Expr::BinaryExpr(binary_expr) => BinaryExpr {
-                op: binary_expr.op,
-                lhs: self.transform(*binary_expr.lhs).into(),
-                rhs: self.transform(*binary_expr.rhs).into(),
-            }
-            .into(),
-            Expr::UnaryExpr(unary_expr) => UnaryExpr {
-                op: unary_expr.op,
-                rhs: self.transform(*unary_expr.rhs).into(),
+    fn transform(&self, target: Rc<Expr>) -> Rc<Expr> {
+        fn fill(cache: &mut HashMap<u64, Rc<Expr>>, t: Rc<Expr>, r: Rc<Expr>) -> Rc<Expr> {
+            Rc::clone(cache.entry(hash(t.as_ref())).or_insert_with(|| r))
+        }
+
+        fn transform(
+            rule: &BuiltRule,
+            target: Rc<Expr>,
+            cache: &mut HashMap<u64 /* ptr address */, Rc<Expr>>,
+        ) -> Rc<Expr> {
```

Comment:
move the comment here, it makes the code less readable.

Reply:
Move it where?

Reply:
just get rid of it.

Reply:
I think it's a valuable comment to explain why we're indexing against an integer though

Reply:
ok, I just think its inconsitent that we have comments in places where things are used like here and in where we define the Hashmap whereas when dealing with the definitions of unbuilt rules, there's only comments in the implementation of UR and not when UR is being used.

Reply:
oh true, removed, forgot that the cache is stored below